Update ERC-6538 Deployment Method and Security Considerations sections (#2)
garyghayrat/ERCs: ERCS/erc-6538.md

  {
    (bool success, bytes memory result) =
      signer.staticcall(abi.encodeCall(IERC1271.isValidSignature, (hash, signature)));
    return (
      success && result.length >= 32
        && abi.decode(result, (bytes32)) == bytes32(IERC1271.isValidSignature.selector)
    );
  }

  /// @dev The following code is from OpenZeppelin's `ECDSA.sol` file. Permalink:
  /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e70a0118ef10773457f670671baefad2c5ea610d/contracts/utils/cryptography/ECDSA.sol
  function tryRecover(bytes32 hash, bytes memory signature)
    internal
    pure
    returns (address, RecoverError, bytes32)
  {
    if (signature.length == 65) {
      bytes32 r;
      bytes32 s;
      uint8 v;
      // ecrecover takes the signature parameters, and the only way to get them
      // currently is to use assembly.
      /// @solidity memory-safe-assembly
      assembly {
        r := mload(add(signature, 0x20))
        s := mload(add(signature, 0x40))
        v := byte(0, mload(add(signature, 0x60)))
      }
      return tryRecover(hash, v, r, s);
    } else {
      return (address(0), RecoverError.InvalidSignatureLength, bytes32(signature.length));
    }
  }

>   /// @notice Overload of {ECDSA-tryRecover} that receives the `v`, `r` and `s` signature fields

Reviewer: Similar comment to the other PR about fixing all natspec to not use OZ's custom style

Author: I'll update this once we merge https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5

The contract is specified below. A one byte integer is used to identify the stealth address scheme. This integer is used to differentiate between different stealth address schemes. A mapping from the scheme ID to it's specification is maintained at [this](../assets/eip-5564/scheme_ids.md) location.

```solidity

> pragma solidity ^0.8.20;

Reviewer: Since we're specifying an exact contract, we should specify an exact compiler version too. Feel free to update this suggestion with whichever version is actually being used
```suggestion
pragma solidity 0.8.23;
```

Author: Update in https://github.com/garyghayrat/ERCs/pull/2/commits/0163a5a14a3b9517802953c5b6c7a5b68903cd4a

/// @notice Registry to map an address or other identifier to its stealth meta-address.
contract ERC5564Registry {
  /// @dev Emitted when a registrant updates their stealth meta-address.
  event StealthMetaAddressSet(
    address indexed registrant, uint256 indexed scheme, bytes stealthMetaAddress
  );

  /// @notice Maps a registrant's identifier to the scheme to the stealth meta-address.
  /// @dev Registrant may be a 160 bit address or other recipient identifier, such as an ENS name.
  /// @dev Scheme is an integer identifier for the stealth address scheme.
  /// @dev MUST return zero if a registrant has not registered keys for the given inputs.
  mapping(address => mapping(uint256 => bytes)) public stealthMetaAddressOf;

>   mapping(address user => uint256 nonce) public nonceOf;
>   bytes32 private constant TYPE_HASH =
>     keccak256("EIP712Domain(string name,string version,uint256 chainId,address registryContract)");

Reviewer: These should have natspec comments, feel free to tweak my suggestions. I also made the typehash public because it may be useful for users.
```suggestion

  /// @notice Next nonce expected from `user` to use when signing for `registerKeysOnBehalf`.
  mapping(address user => uint256 nonce) public nonceOf;
  
  /// @dev EIP-712 Type hash used in `registerKeysOnBehalf`
  bytes32 public constant TYPE_HASH =
    keccak256("EIP712Domain(string name,string version,uint256 chainId,address registryContract)");
```

Author: Updated in the same commit as above:  https://github.com/garyghayrat/ERCs/pull/2/commits/0163a5a14a3b9517802953c5b6c7a5b68903cd4a

  /// @param scheme An integer identifier for the stealth address scheme.
  /// @param stealthMetaAddress The stealth meta-address to register.
  function registerKeys(uint256 scheme, bytes memory stealthMetaAddress) external {
    stealthMetaAddressOf[msg.sender][scheme] = stealthMetaAddress;
  }

  /// @notice Sets the `registrant`s stealth meta-address for the given scheme.
  /// @param registrant Recipient identifier, such as an ENS name.
  /// @param schemeId An integer identifier for the stealth address scheme.
  /// @param signature A signature from the `registrant` authorizing the registration.
  /// @param stealthMetaAddress The stealth meta-address to register.
  /// @dev MUST support both EOA signatures and EIP-1271 signatures.
  /// @dev MUST revert if the signature is invalid.
  function registerKeysOnBehalf(
    address registrant,
    uint256 schemeId,
    bytes memory signature,
    bytes memory stealthMetaAddress
  ) external {
    bytes32 digest = _hashTypedDataV4(
      keccak256(
        abi.encode(TYPE_HASH, registrant, schemeId, stealthMetaAddress, nonceOf[registrant]++)
      )
    );

>     require(isValidSignatureNow(registrant, digest, signature), "Invalid signature");

Reviewer: This method is not defined anywhere in this contract in the ERC, but the ERC should include the full source code.

Author: Adde the full source code in the same commit as above:  https://github.com/garyghayrat/ERCs/pull/2/commits/0163a5a14a3b9517802953c5b6c7a5b68903cd4a

    stealthMetaAddressOf[registrant][schemeId] = stealthMetaAddress;
    emit StealthMetaAddressSet(registrant, schemeId, stealthMetaAddress);
  }
}

> ```

Reviewer: Since we have a nonce, we also need a method for users to bump their nonce manually so they can invalidate existing signatures.

```solidity
function incrementNonce() external {
  noncesOf[msg.sender]++;
}

Author: Added in the same commit as above:  https://github.com/garyghayrat/ERCs/pull/2/commits/0163a5a14a3b9517802953c5b6c7a5b68903cd4a

### Deployment Method

> This contract is going to be deployed using the CREATE2 method. A predetermined address will use Nick's deterministic deployment proxy address of 0x4e59b44847b379578588920ca78fbf26c0b4956c to deploy the contract.
> An empty string, "", is going to be used as the salt.

Reviewer: Salt must be 32 bytes so an empty string is not valid. You can use all zeros, but we may want to mine a salt for a fun vanity address
```suggestion
This contract is going to be deployed using CREATE2 via the deterministic deployer at `0x4e59b44847b379578588920ca78fbf26c0b4956c` with a salt of TODO.
```

Author: Updated in the same commit as above:  https://github.com/garyghayrat/ERCs/pull/2/commits/0163a5a14a3b9517802953c5b6c7a5b68903cd4a

## Security Considerations

> In the event of a compromised private key, the registrant should promptly un-register from the stealth key registry to prevent loss of future funds sent to the compromised account.

Reviewer: We should go into a bit more detail here because this is easier said than done—once compromised, any ETH sent gets drained by bots right away, so they have to do this with a signature and relayer, or use a private mempool like flashbots to batch the transactions if they chain supports it.

Author: Tried expanding in https://github.com/garyghayrat/ERCs/pull/2/commits/9d290de39729188aedaf951150e79e474594c907 but still some room for improvement?

Reviewer: That works, though it sounds like the signature must go through a private mempool, so maybe tweak it to clarify that approach is a separate solution than the sig solution

Author: You're right, how about this: https://github.com/garyghayrat/ERCs/pull/2/commits/b7d9a58dfdc90095c2069b60d0cca557419e322d
